### PR TITLE
Interpolation upgrade

### PR DESCRIPTION
**Pull request by 1killerbehindyou1**
- doadano klase odpowiedzialną za wyczytywanie i wyświetlanie pliku - ImageWiewer 
- dodano klase DataMAnager do zarządzania danymi w pamiei programu
- dodano klase File manager do obsługiwania operacji zapisu i odczytu z pliku
- naprawiono problem ze złym obliczaniem koloru 
- zaimplenemtowano transformacje prostokąta pixel po pixelu, przebudowano klasę Interpolator

### CODE REVIEW

File: PointsInPolygon/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +25,19 @@ set(PROJECT_SOURCES
...
+    # variable in the caller scope but in the parent scope of the block()
+    set(${RESULT_VARIABLE} "new-value")
+    message("MY_VAR is equal: " ${RESULT_VARIABLE})
+    #return(PROPAGATE ${RESULT_VARIABLE})
+  
+endfunction()
+
+
+set(MY_VAR "initial-value")
+multi_scopes(MY_VAR)
```
Comment (Nutek) [inline PointsInPolygon/CMakeLists.txt:40]: Is it PROD version?

**Review by Nutek — CHANGES_REQUESTED** [review]

File: BasicGUI/FileManager.cpp [hunk 2]
```diff
@@ -0,0 +1,9 @@
+
+#include "FileManager.h"
+#include <iostream>
+#include <QDebug>
+
+
+void FileManager::loadFromFile(){}
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/FileManager.cpp:7]: Resolved by remowing this file from project

File: BasicGUI/DataManager.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#include "DataManager.h"
+#include "LedRuler.h"
+#include <iostream>
+
+DataManager::DataManager(QObject* parent) : QObject(parent){
+
+}
+
+void DataManager::loadFromFile(){
+
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/DataManager.cpp:11]: Resolved by remowing this file from project

File: BasicGUI/Interpolator.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        {
+             current_point.point.setX(rect_top_left.x() + xx);
+            
+            current_point.calcAngle();
+            current_point.calcModule();
+            current_point.updateAngle(ext_angle);
+            current_point.updatePoint(center_of_rot);
+            vector_points.push_front(current_point.point);
+
+            std::cout << " point X: " << current_point.point.x() << " point Y:  "<< current_point.point.y() << std::endl;
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.cpp:74]: Removed this line from project.

File: BasicGUI/Interpolator.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    QPointF point;
+    float convertAngleToRad(float ext_angle);
+private:
+    
+
+
+};
+
+
+struct Component{
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.h:32]: Class name changed from Component to ColorComponent.

File: CMakeLists.txt [hunk 6]
```diff
@@ -8,3 +8,7 @@ add_subdirectory(BasicGUI)
 
 add_subdirectory(PointsInPolygon)
 
+message("CMAKE GENERATOR: " ${CMAKE_GENERATOR})
+message("CMAKE_MAKE_PROGRAM:  " ${CMAKE_MAKE_PROGRAM})
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply CMakeLists.txt:12]: Remove this lines from file CMakeLisis.txt

File: PointsInPolygon/CMakeLists.txt (hunk 1 above)
Comment (1killerbehindyou1) [reply PointsInPolygon/CMakeLists.txt:40]: Remove this lines from file CMakeLisis.txt

File: BasicGUI/Interpolator.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    QColor interpolatorSetLedColor(QVector<QPointF> vector_points);
+    QVector<QPointF> interpolatorTransform(QPoint sr_obr, QRect rect, float angle);
+
+    
+
+    QPixmap inerpolator_pixmap;
+
+private:
+    Component red_component;
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.h:54]: Class name changed from Component to ColorComponent.

File: BasicGUI/LedRuler.cpp [hunk 8]
```diff
@@ -1,143 +1,69 @@
...
+    m_rotation = 90;
+    m_spacing = 5;
+    m_size =3;
 }
 
 void LedRuler::paint(QPainter *painter)
-{
-    painter->translate(m_point);
+{   
+    DataManager::m_projection = new (QVector<QVector<QColor>>);
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/LedRuler.cpp:22]: Resolved by making m_projection = new (QVector<QVector<QColor>>) automatic variable.

File: BasicGUI/Interpolator.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            current_point.point.setY(rect_top_left.y() + yy);
+            current_point.point.setX(rect_top_left.x() + xx); 
+        }
+    }
+    return vector_points;
+} 
+
+float PointInReferenceSystem::convertAngleToRad(float ext_angle) {return ext_angle * 3.14159 / 180; }
+
+void PointInReferenceSystem::calcAngle() { angle = atan2(point.y(), point.x()); }
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.cpp:84]: ok

File: BasicGUI/Interpolator.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+         
+        for (int xx = 0; xx < rect.width(); xx++)
+        {
+             current_point.point.setX(rect_top_left.x() + xx);
+            
+            current_point.calcAngle();
+            current_point.calcModule();
+            current_point.updateAngle(ext_angle);
+            current_point.updatePoint(center_of_rot);
+            vector_points.push_front(current_point.point);
```
[the remark this answers is not in the document]
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.cpp:104]: Implemented

File: BasicGUI/LedRuler.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Nutek) [reply BasicGUI/LedRuler.cpp]: As discussed offline: please double check if this file is up to date in PR.

File: BasicGUI/Interpolator.cpp [hunk 11]
```diff
@@ -71,41 +44,38 @@ QColor Interpolator::interpolatorSetLedColor(QVector<QPointF> vector_points)
...
-
-QVector<QPointF> Interpolator::interpolatorTransform(QPoint center_of_rot,
-                                                     QRect rect,
-                                                     float deg_angle)
+//////////////////////////////////////////////////////////////////
+QVector<QPointF> Interpolator::interpolatorTransform(QPoint a_center_of_rot,
+                                                     QRect a_rect,
+                                                     float a_deg_angle)
 {
     QVector<QPointF> vector_points;
```
[the remark this answers is not in the document]
Comment (Nutek) [reply BasicGUI/Interpolator.cpp:52]: `QPolygon` improvemnt will be a part of next exercise with benchmarking. Please keep `QVector<QPointF> for now.`

File: BasicGUI/Interpolator.cpp [hunk 12]
```diff
@@ -71,41 +44,38 @@ QColor Interpolator::interpolatorSetLedColor(QVector<QPointF> vector_points)
...
- *       rectangle
- */
-
-QVector<QPointF> Interpolator::interpolatorTransform(QPoint center_of_rot,
-                                                     QRect rect,
-                                                     float deg_angle)
+//////////////////////////////////////////////////////////////////
+QVector<QPointF> Interpolator::interpolatorTransform(QPoint a_center_of_rot,
+                                                     QRect a_rect,
+                                                     float a_deg_angle)
```
Comment (Nutek) [inline BasicGUI/Interpolator.cpp:50]: Please consider extracting math to separate object like `Rotate` and provide it by base class like `Transform` where `operator()(QPointF)` will be responsible to compute transformed point.

Comment (Nutek) [reply BasicGUI/Interpolator.cpp:50]: As an alternative you could pass `Rotate` by `std::function<QPointF(QPointF)>`

File: BasicGUI/Interpolator.cpp [hunk 13]
```diff
@@ -71,41 +44,38 @@ QColor Interpolator::interpolatorSetLedColor(QVector<QPointF> vector_points)
...
+            t_angle = Transform::calcAngle(t_current_point);
+            t_module = Transform::calcModule(t_current_point, t_angle);
+            t_angle = Transform::updateAngle(t_ext_angle, t_angle);
+            t_current_point = Transform::updatePoint(
+                a_center_of_rot, t_current_point, t_module, t_angle);
+            vector_points.push_front(t_current_point);
+
+            t_angle = 0;
+            t_current_point.setY(t_rect_top_left.y() + yy);
+            t_current_point.setX(t_rect_top_left.x() + xx);
```
Comment (Nutek) [inline BasicGUI/Interpolator.cpp:78]: This math should be extracted

File: BasicGUI/Interpolator.cpp [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    for (int y = 0; y < rect.height(); y++)
+    {
+        curr_point.setY(rect.topLeft().y() + y);
+
+        for (int x = 0; x < rect.width(); x++)
+        {
+            curr_point.setX(rect.topLeft().x() + x);
+
+            Transform trans(center_of_rot, deg_angle);
+            vector_points.push_front(trans(curr_point));
```
Comment (Nutek) [inline BasicGUI/Interpolator.cpp:65]: This is what I've looked for. Is that looks better?

File: BasicGUI/Interpolator.cpp [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+    for (int y = 0; y < rect.height(); y++)
+    {
+        curr_point.setY(rect.topLeft().y() + y);
+
+        for (int x = 0; x < rect.width(); x++)
+        {
+            curr_point.setX(rect.topLeft().x() + x);
+
+            Transform trans(center_of_rot, deg_angle);
```
Comment (Nutek) [inline BasicGUI/Interpolator.cpp:64]: This object depends on variables which are constant in the loop. Please consider moving it outside of loop.

File: BasicGUI/Interpolator.h [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+//////////////////////////////////////////////////////////////////
+class Interpolator : public QObject
+{
+    Q_OBJECT
+
+public:
+    Interpolator(QObject* parent = 0);
+
+    QColor interpolatorSetLedColor(QVector<QPointF> vector_points);
+    QVector<QPointF> interpolatorTransform(QPoint sr_obr, QRect rect,
```
Comment (Nutek) [inline BasicGUI/Interpolator.h:32]: `Transform` could be a parameter, but we can leave for now.

Name of parameters are inconsistent between declaration and definition of method.

File: BasicGUI/Transformation.h [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+public:
+    Transform(QPoint rotCenter, float deg_angle);
+    QPointF transformFrom(QPointF point);
+    QPointF operator()(QPointF point);
+
+private:
+    float static calcModule(QPointF point, float angle);
+    float static calcAngle(QPointF point);
+    float static convertAngleToRad(float deg_angle);
+    float static updateAngle(float ext_angle, float angle);
```
Comment (Nutek) [inline BasicGUI/Transformation.h:22]: Those methods are static and private. There is no benefit they are part of class.

Please move them to anonymous namespace in source file - still you will be able to use them as implementation details, and API of your class will be cleaner.

**Comment by Nutek** [comment]
It's not perfect, but may be for now. Good state for benchmark.

**Review by Nutek — APPROVED** [review]

File: BasicGUI/Interpolator.cpp (hunk 14 above)
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.cpp:65]: Definetly more transparent :)

File: BasicGUI/Interpolator.cpp (hunk 15 above)
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.cpp:64]: line 64: Transform trans(center_of_rot, deg_angle);
pulled out of the loop.

File: BasicGUI/Interpolator.h (hunk 16 above)
Comment (1killerbehindyou1) [reply BasicGUI/Interpolator.h:32]: fixed

File: BasicGUI/Transformation.h (hunk 17 above)
Comment (1killerbehindyou1) [reply BasicGUI/Transformation.h:22]: Implemented

**Comment by 1killerbehindyou1** [comment]
Refactoring and updating the color interpolation mechanism and drawing ruler 07.12.2022

**Pull request merged**